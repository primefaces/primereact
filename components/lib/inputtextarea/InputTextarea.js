import * as React from 'react';
import { Tooltip } from '../tooltip/Tooltip';
import { classNames, DomHandler, ObjectUtils } from '../utils/Utils';

export const InputTextarea = React.memo(React.forwardRef((props, ref) => {
    const elementRef = React.useRef(ref);
    const cachedScrollHeight = React.useRef(0);

    const onFocus = (event) => {
        if (props.autoResize) {
            resize();
        }

        props.onFocus && props.onFocus(event);
    }

    const onBlur = (event) => {
        if (props.autoResize) {
            resize();
        }

        props.onBlur && props.onBlur(event);
    }

    const onKeyUp = (event) => {
        if (props.autoResize) {
            resize();
        }

        props.onKeyUp && props.onKeyUp(event);
    }

    const onInput = (event) => {
        if (props.autoResize) {
            resize();
        }

        props.onInput && props.onInput(event);

        const target = event.target;
        ObjectUtils.isNotEmpty(target.value) ? DomHandler.addClass(target, 'p-filled') : DomHandler.removeClass(target, 'p-filled');
    }

    const resize = (initial) => {
        const inputEl = elementRef.current;

        if (inputEl && DomHandler.isVisible(inputEl)) {
            if (!cachedScrollHeight.current) {
                cachedScrollHeight.current = inputEl.scrollHeight;
                inputEl.style.overflow = 'hidden';
            }

            if (cachedScrollHeight.current !== inputEl.scrollHeight || initial) {
                inputEl.style.height = '';
                inputEl.style.height = inputEl.scrollHeight + 'px';

                if (parseFloat(inputEl.style.height) >= parseFloat(inputEl.style.maxHeight)) {
                    inputEl.style.overflowY = 'scroll';
                    inputEl.style.height = inputEl.style.maxHeight;
                }
                else {
                    inputEl.style.overflow = 'hidden';
                }

                cachedScrollHeight.current = inputEl.scrollHeight;
            }
        }
    }

    const isFilled = React.useMemo(() => (
        ObjectUtils.isNotEmpty(props.value) || ObjectUtils.isNotEmpty(props.defaultValue) || (elementRef.current && ObjectUtils.isNotEmpty(elementRef.current.value))
    ), [props.value, props.defaultValue]);

    React.useEffect(() => {
        ObjectUtils.combinedRefs(elementRef, ref);
    }, [elementRef, ref]);

    React.useEffect(() => {
        if (props.autoResize) {
            resize(true);
        }
    }, [props.autoResize]);

    const hasTooltip = ObjectUtils.isNotEmpty(props.tooltip);
    const otherProps = ObjectUtils.findDiffKeys(props, InputTextarea.defaultProps);
    const className = classNames('p-inputtextarea p-inputtext p-component', {
        'p-disabled': props.disabled,
        'p-filled': isFilled,
        'p-inputtextarea-resizable': props.autoResize
    }, props.className);

    return (
        <>
            <textarea ref={elementRef} {...otherProps} className={className} onFocus={onFocus} onBlur={onBlur} onKeyUp={onKeyUp} onInput={onInput}></textarea>
            {hasTooltip && <Tooltip target={elementRef} content={props.tooltip} {...props.tooltipOptions} />}
        </>
    )
}));

InputTextarea.displayName = 'InputTextarea';
InputTextarea.defaultProps = {
    __TYPE: 'InputTextarea',
    autoResize: false,
    tooltip: null,
    tooltipOptions: null,
    onInput: null
}
