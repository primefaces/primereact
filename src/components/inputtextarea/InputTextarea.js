import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export class InputTextarea extends Component {

    static defaultProps = {
        autoResize: false,
        onInput: null,
        onClick: null
    };

    static propTypes = {
        autoResize: PropTypes.bool,
        onInput: PropTypes.func,
        onClick: PropTypes.func
    };
    
    constructor(props) {
        super(props);
        this.onFocus = this.onFocus.bind(this);
        this.onBlur = this.onBlur.bind(this);
        this.onKeyUp = this.onKeyUp.bind(this);
        this.onInput = this.onInput.bind(this);
        this.onClick = this.onClick.bind(this);
        this.state = {filled: false};

        this.textareaProps = Object.assign({}, this.props);
        delete this.textareaProps.autoResize;
        delete this.textareaProps.onInput;
    }

    onFocus(e) {
        if(this.props.autoResize) {
            this.resize();
        }
    }

    onBlur(e) {
        if(this.props.autoResize) {
            this.resize();
        }
    }

    onKeyUp(e) {
        if(this.props.autoResize) {
            this.resize();
        }
    }

  onClick(e) {
    if(this.props.onClick) {
      this.props.onClick();
    } else {
      e.stopPropagation();
    }
  }

    resize () {
        let linesCount = 0,
        lines = this.textareaElement.value.split('\n');

        for(let i = lines.length-1; i >= 0 ; --i) {
            linesCount += Math.floor((lines[i].length / parseInt(this.props.cols, 10)) + 1);
        }

        this.textareaElement.rows = (linesCount >= parseInt(this.props.rows, 10)) ? (linesCount + 1) : parseInt(this.props.rows, 10);
    }

    onInput(e) {
        if(this.props.onInput) {
            this.props.onInput();
        }

        this.updateFilledState(e);
    }

    updateFilledState(e) {
        let _filled = (e.target.value && e.target.value.length) ? true : false;
        this.setState({filled: _filled});
    }
    
    componentDidMount() {
        let _value =  this.textareaProps.value||this.textareaProps.defaultValue,
        _filled = (_value && _value.length) ? true : false;
        
        this.setState({filled: _filled});
    }

    componentWillReceiveProps(nextProps) {
        if(nextProps.hasOwnProperty('value')) {
            this.textareaProps.value = nextProps.value;
        }
    }   

    render() {
        var className = classNames('ui-inputtext ui-corner-all ui-state-default ui-widget', this.props.className, {
            'ui-state-disabled': this.props.disabled,
            'ui-state-filled': this.state.filled
        });

        return <textarea {...this.textareaProps} className={className} ref={(input) => {this.textareaElement = input;}} onFocus={this.onFocus} onBlur={this.onBlur} onKeyUp={this.onKeyUp} onInput={this.onInput} onClick={this.onClick}></textarea>;
    }
}