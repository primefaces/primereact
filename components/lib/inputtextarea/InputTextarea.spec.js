import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';
import { InputTextarea } from './InputTextarea';

describe('InputTextarea', () => {
    test('when textarea is enabled then textarea accepts data entry and have filled state', async () => {
        // Arrange
        const { container } = render(<InputTextarea />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        await userEvent.type(input, 'abc');

        // Act
        expect(input).toBeEnabled();
        expect(input).toHaveValue('abc');
        expect(container).toMatchSnapshot();
    });
    test('when textarea is is autosizing and has rows and columns it is rendered correctly', () => {
        // Arrange
        const blurOn = jest.fn();
        const { container } = render(<InputTextarea autoResize rows={3} cols={12} onBlur={blurOn} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        fireEvent.blur(input);

        // Act and Assert
        expect(blurOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot();
    });
    test('when input is blank it should not have filled state', async () => {
        // Arrange
        const { container } = render(<InputTextarea disabled={false} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        fireEvent.input(input, { target: { value: '' } });

        // Act
        expect(input).toBeEnabled();
        expect(container).toMatchSnapshot();
    });
    test('when input is is set for validation only', async () => {
        // Arrange
        const { container } = render(<InputTextarea validateOnly keyfilter={`alpha`} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        await userEvent.type(input, 'def');

        // Act
        expect(container).toMatchSnapshot();
    });
    test('when input is disabled it should render as disabled', async () => {
        // Arrange
        const { container } = render(<InputTextarea disabled />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        await userEvent.type(input, '123');

        // Act
        expect(input).toBeDisabled();
        expect(container).toMatchSnapshot();
    });
    test('when input is using keyfilter for integers accept integer input', async () => {
        // Arrange
        const keydownOn = jest.fn();
        const keyupOn = jest.fn();
        const { container } = render(<InputTextarea keyfilter="int" onKeyDown={keydownOn} onKeyUp={keyupOn} autoResize />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        await userEvent.type(input, '123');

        // Act
        expect(input).toHaveValue('123');
        expect(keydownOn).toHaveBeenCalledTimes(3);
        expect(keyupOn).toHaveBeenCalledTimes(3);
    });
    test('when input is using keyfilter for integers do not accept alphabetic input', async () => {
        // Arrange
        const keydownOn = jest.fn();
        const { container } = render(<InputTextarea keyfilter="int" onKeyDown={keydownOn} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        await userEvent.type(input, 'abc');

        // Act
        expect(input).toHaveValue('');
        expect(keydownOn).toHaveBeenCalledTimes(3);
    });
    test('when input is using keyfilter for alphabetic accept paste of alphabetic values', async () => {
        // Arrange
        const pasteOn = jest.fn();
        const { container } = render(<InputTextarea keyfilter="alpha" onPaste={pasteOn} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        input.focus();
        await userEvent.paste('abc');

        // Act
        expect(input).toHaveValue('abc');
        expect(pasteOn).toHaveBeenCalledTimes(1);
    });
    test('when input is using keyfilter for alphabetic do not accept paste of integer values', async () => {
        // Arrange
        const pasteOn = jest.fn();
        const { container } = render(<InputTextarea keyfilter="alpha" onPaste={pasteOn} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        input.focus();
        await userEvent.paste('123');

        // Act
        expect(input).toHaveValue('');
        expect(pasteOn).toHaveBeenCalledTimes(1);
    });
    test('when input is using onInput make sure it is called', async () => {
        // Arrange
        const inputOn = jest.fn();
        const ref = React.createRef();
        const { container } = render(<InputTextarea ref={ref} onInput={inputOn} />);
        const input = container.getElementsByTagName('textarea')[0];

        // Act
        ref.current.value = 'Jest testing value';
        input.focus();
        await userEvent.paste(' abc');

        // Assert
        expect(input).toHaveValue('Jest testing value abc');
        expect(inputOn).toHaveBeenCalledTimes(1);
    });
});
