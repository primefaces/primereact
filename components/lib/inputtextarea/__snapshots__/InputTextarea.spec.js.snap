// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextarea when input is blank it should not have filled state 1`] = `
<div>
  <textarea
    class="p-inputtextarea p-inputtext p-component"
  />
</div>
`;

exports[`InputTextarea when input is disabled it should render as disabled 1`] = `
<div>
  <textarea
    class="p-inputtextarea p-inputtext p-component p-disabled"
    disabled=""
  />
</div>
`;

exports[`InputTextarea when input is is set for validation only 1`] = `
<div>
  <textarea
    class="p-inputtextarea p-inputtext p-component p-filled"
  />
</div>
`;

exports[`InputTextarea when textarea is enabled then textarea accepts data entry and have filled state 1`] = `
<div>
  <textarea
    class="p-inputtextarea p-inputtext p-component p-filled"
  />
</div>
`;

exports[`InputTextarea when textarea is is autosizing and has rows and columns it is rendered correctly 1`] = `
<div>
  <textarea
    class="p-inputtextarea p-inputtext p-component p-inputtextarea-resizable"
    cols="12"
    rows="3"
    style="overflow: hidden; height: 0px;"
  />
</div>
`;
